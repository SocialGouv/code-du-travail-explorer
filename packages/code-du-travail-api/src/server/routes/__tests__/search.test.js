const request = require("supertest");
const Koa = require("koa");
const winston = require("winston");
const { SOURCES } = require("@socialgouv/cdtn-sources");
const router = require("../search");

const getSearchBody = require("../search/search.elastic");
const getSemBody = require("../search/search.sem");
const getDocumentByUrlQuery = require("../search/getDocumentByUrlQuery");

const { logger } = require("../../utils/logger");
require("./__mocking__/mockNLP");

jest.mock("@cdt/data/indexing/vectorizer");

logger.level = winston.error;

const app = new Koa();
app.use(router.routes());

it("asks same sources wether it is search sem or search elastic and gets a description", () => {
  const searchBody = getSearchBody({ sources: [SOURCES.CDT] });
  const semBody = getSemBody({ sources: [SOURCES.CDT] });
  const getEsRefBody = getDocumentByUrlQuery("/code-du-travail/slug");
  expect(searchBody._source).toEqual(semBody._source);
  expect(getEsRefBody._source).toEqual(semBody._source);
  expect(searchBody._source).toContain("description");
});

it("returns search results for demission from datafiller", async () => {
  const response = await request(app.callback()).get(
    "/api/v1/search?q=démission"
  );
  expect(response.status).toBe(200);
  expect(response.body).toMatchSnapshot(); // datafiller is involved here (no ES or Sem)
});

it("returns 3 search results for demission from elastic if size = 3", async () => {
  const response = await request(app.callback()).get(
    "/api/v1/search?q=démission&skipSavedResults&size=3"
  );
  expect(response.status).toBe(200);
  expect(response.body.documents.length).toBe(3);
});

it("returns search results for demission from elastic", async () => {
  const response = await request(app.callback()).get(
    "/api/v1/search?q=démission&skipSavedResults"
  );
  expect(response.status).toBe(200);
  expect(response.body).toMatchSnapshot();
});

it("returns article results when searching article R1225-18", async () => {
  const response = await request(app.callback()).get(
    `/api/v1/search?q=r1225-18`
  );
  expect(response.status).toBe(200);
  expect(response.body).toMatchSnapshot();
});
