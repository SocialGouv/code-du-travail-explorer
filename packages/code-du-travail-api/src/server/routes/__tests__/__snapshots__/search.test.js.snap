// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`return conventions 1`] = `
Object {
  "articles": Array [],
  "documents": Array [
    Object {
      "_score": 20.0325,
      "algo": "fulltext",
      "breadcrumbs": Array [],
      "cdtnId": "16752da066",
      "description": "Idcc 1747 : Activités industrielles de boulangerie et pâtisserie",
      "slug": "1747-activites-industrielles-de-boulangerie-et-patisserie",
      "source": "conventions_collectives",
      "title": "Convention collective nationale des activités industrielles de boulangerie et pâtisserie du 13 juillet 1993. Mise à jour par avenant n°10 du 11 octobre 2011.",
      "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635691",
    },
  ],
  "themes": Array [],
}
`;

exports[`returns article results when searching article R1225-18 1`] = `
Object {
  "articles": Array [
    Object {
      "_score": 22.683619,
      "cdtnId": "f13147ec33",
      "description": "Le salarié informe son employeur de sa démission, en application de l'article L. 1225-66 , par lettre recommandée avec avis de réception ou remise contre récépissé. Il adresse à l'employeur sa demande de réembauche, en application de l'article L. 1225-67 , par lettre recommandée avec avis de réception ou remise contre récépissé",
      "slug": "r1225-18",
      "source": "code_du_travail",
      "title": "R1225-18",
      "url": "https://www.legifrance.gouv.fr/affichCodeArticle.do;?idArticle=LEGIARTI000018537778&cidTexte=LEGITEXT000006072050",
    },
  ],
  "documents": Array [],
  "themes": Array [],
}
`;

exports[`returns search results for demission from elastic 1`] = `
Object {
  "articles": Array [
    Object {
      "_score": 3.589242,
      "cdtnId": "f13147ec33",
      "description": "Le salarié informe son employeur de sa démission, en application de l'article L. 1225-66 , par lettre recommandée avec avis de réception ou remise contre récépissé. Il adresse à l'employeur sa demande de réembauche, en application de l'article L. 1225-67 , par lettre recommandée avec avis de réception ou remise contre récépissé",
      "slug": "r1225-18",
      "source": "code_du_travail",
      "title": "R1225-18",
      "url": "https://www.legifrance.gouv.fr/affichCodeArticle.do;?idArticle=LEGIARTI000018537778&cidTexte=LEGITEXT000006072050",
    },
    Object {
      "_score": 3.4476256,
      "cdtnId": "ea03713a82",
      "description": "En cas de démission, l'existence et la durée du préavis sont fixées par la loi, ou par convention ou accord collectif de travail. En l'absence de dispositions légales, de convention ou accord collectif de travail relatifs au préavis, son existence et sa durée résultent des usages pratiqués dans la localité et dans la profession. Un décret en Conseil d'Etat détermine les modalités d'application du présent article",
      "slug": "l1237-1",
      "source": "code_du_travail",
      "title": "L1237-1",
      "url": "https://www.legifrance.gouv.fr/affichCodeArticle.do;?idArticle=LEGIARTI000006901174&cidTexte=LEGITEXT000006072050",
    },
  ],
  "documents": Array [
    Object {
      "_score": 39.38178,
      "algo": "both",
      "breadcrumbs": Array [
        Object {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        Object {
          "label": "Démission",
          "position": 1,
          "slug": "/themes/demission",
        },
      ],
      "cdtnId": "8c42435f9a",
      "description": "La démission permet au salarié, à son initiative, de rompre son contrat de travail, sous conditions.",
      "slug": "demission-dun-salarie",
      "source": "fiches_service_public",
      "title": "Démission d'un salarié",
      "url": "https://www.service-public.fr/particuliers/vosdroits/F2883",
    },
    Object {
      "_score": 1.3151006,
      "action": "Calculer",
      "algo": "both",
      "breadcrumbs": Array [
        Object {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        Object {
          "label": "Démission",
          "position": 1,
          "slug": "/themes/demission",
        },
      ],
      "cdtnId": "1eea193273",
      "description": "Calculez la durée de préavis à respecter en cas de démission",
      "slug": "preavis-demission",
      "source": "outils",
      "title": "Préavis de démission",
    },
    Object {
      "_score": 1.2952552,
      "action": "Estimer",
      "algo": "semantic",
      "breadcrumbs": Array [
        Object {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        Object {
          "label": "Licenciement : droits des salariés et procédures",
          "position": 4,
          "slug": "/themes/licenciement-droits-des-salaries-et-procedures",
        },
      ],
      "cdtnId": "d8a3605790",
      "description": "Estimez simplement le montant de l’indemnité de licenciement",
      "slug": "indemnite-licenciement",
      "source": "outils",
      "title": "Indemnité de licenciement",
    },
    Object {
      "_score": 37.928764,
      "algo": "both",
      "breadcrumbs": Array [
        Object {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        Object {
          "label": "Démission",
          "position": 1,
          "slug": "/themes/demission",
        },
      ],
      "cdtnId": "38474d168c",
      "description": "L'assistante maternelle (ou assistant maternel) agréée qui souhaite rompre son contrat de travail doit informer le particulier employeur par écrit et respecter un préavis. À la fin de celui-ci, l'employeur remet les documents de fin de contrat.",
      "slug": "demission-dune-assistante-maternelle",
      "source": "fiches_service_public",
      "title": "Démission d'une assistante maternelle",
      "url": "https://www.service-public.fr/particuliers/vosdroits/F33164",
    },
    Object {
      "_score": 32.778652,
      "algo": "both",
      "breadcrumbs": Array [
        Object {
          "label": "Santé, sécurité et conditions de travail",
          "position": 6,
          "slug": "/themes/sante-securite-et-conditions-de-travail",
        },
        Object {
          "label": "Santé au travail",
          "position": 3,
          "slug": "/themes/sante-au-travail",
        },
        Object {
          "label": "Maladie",
          "position": 1,
          "slug": "/themes/maladie",
        },
      ],
      "cdtnId": "36c2fc8868",
      "description": "Le salarié en arrêt maladie pendant son préavis perçoit les indemnités habituelles. Le préavis est prolongé si l'arrêt est d'origine professionnelle.",
      "slug": "arret-maladie-pendant-le-preavis-quelles-consequences",
      "source": "fiches_service_public",
      "title": "Arrêt maladie pendant le préavis : quelles conséquences ?",
      "url": "https://www.service-public.fr/particuliers/vosdroits/F2614",
    },
    Object {
      "_score": 32.00379,
      "algo": "both",
      "breadcrumbs": Array [
        Object {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        Object {
          "label": "Documents à remettre au salarié",
          "position": 0,
          "slug": "/themes/documents-a-remettre-au-salarie",
        },
      ],
      "cdtnId": "59a4d4173c",
      "description": "Le certificat de travail est un document remis par l'employeur au salarié à la date de fin du contrat de travail. Il est remis au salarié pour tout type de contrat de travail (CDI, CDD...), quel que soit le motif de la rupture (démission, licenciement...).",
      "slug": "certificat-de-travail",
      "source": "fiches_service_public",
      "title": "Certificat de travail",
      "url": "https://www.service-public.fr/particuliers/vosdroits/F87",
    },
    Object {
      "_score": 1.2357258,
      "algo": "semantic",
      "breadcrumbs": Array [
        Object {
          "label": "Embauche et contrat de travail",
          "position": 1,
          "slug": "/themes/embauche-et-contrat-de-travail",
        },
        Object {
          "label": "Embauche",
          "position": 1,
          "slug": "/themes/embauche",
        },
        Object {
          "label": "Période d’essai",
          "position": 3,
          "slug": "/themes/periode-dessai",
        },
      ],
      "cdtnId": "772cb955ce",
      "description": "Pendant la période d’essai, le contrat de travail peut être rompu librement par l’employeur. L’employeur doit dans ce cas informer le salarié et respecter un délai de prévenance.",
      "slug": "rupture-de-periode-dessai-a-linitiative-de-lemployeur",
      "source": "modeles_de_courriers",
      "title": "Rupture de période d’essai à l’initiative de l’employeur",
    },
    Object {
      "_score": 1.1560779,
      "algo": "semantic",
      "breadcrumbs": Array [
        Object {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        Object {
          "label": "Rupture conventionnelle",
          "position": 3,
          "slug": "/themes/rupture-conventionnelle",
        },
        Object {
          "label": "Rupture conventionnelle individuelle",
          "position": 1,
          "slug": "/themes/rupture-conventionnelle-individuelle",
        },
      ],
      "cdtnId": "a712ecb976",
      "description": "La rupture conventionnelle individuelle est une modalité de rupture spécifique du CDI. Elle nécessite le consentement de l’employeur et du salarié, et son homologation par l’administration. La rupture ouvre droit à une indemnité de rupture conventionnelle. Ce modèle permet d’initier la procédure de rupture par l’invitation à un premier entretien.",
      "slug": "demande-de-rendez-vous-en-vue-dune-rupture-conventionnelle",
      "source": "modeles_de_courriers",
      "title": "Demande de rendez-vous en vue d’une rupture conventionnelle",
    },
    Object {
      "_score": 1.1512249,
      "algo": "semantic",
      "breadcrumbs": Array [
        Object {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        Object {
          "label": "Documents à remettre au salarié",
          "position": 0,
          "slug": "/themes/documents-a-remettre-au-salarie",
        },
      ],
      "cdtnId": "19837df114",
      "description": "Le certificat de travail est un document obligatoire remis par l’employeur au salarié à la date de fin du contrat de travail.",
      "slug": "certificat-de-travail",
      "source": "modeles_de_courriers",
      "title": "Certificat de travail",
    },
  ],
  "themes": Array [
    Object {
      "_score": 1.9676356,
      "algo": "fulltext",
      "cdtnId": "a9ccf1d613",
      "slug": "demission",
      "source": "themes",
      "title": "Démission",
    },
  ],
}
`;

exports[`returns search results for demission from prequalified requests 1`] = `
Object {
  "articles": Array [
    Object {
      "_score": 3.589242,
      "cdtnId": "f13147ec33",
      "description": "Le salarié informe son employeur de sa démission, en application de l'article L. 1225-66 , par lettre recommandée avec avis de réception ou remise contre récépissé. Il adresse à l'employeur sa demande de réembauche, en application de l'article L. 1225-67 , par lettre recommandée avec avis de réception ou remise contre récépissé",
      "slug": "r1225-18",
      "source": "code_du_travail",
      "title": "R1225-18",
      "url": "https://www.legifrance.gouv.fr/affichCodeArticle.do;?idArticle=LEGIARTI000018537778&cidTexte=LEGITEXT000006072050",
    },
    Object {
      "_score": 3.4476256,
      "cdtnId": "ea03713a82",
      "description": "En cas de démission, l'existence et la durée du préavis sont fixées par la loi, ou par convention ou accord collectif de travail. En l'absence de dispositions légales, de convention ou accord collectif de travail relatifs au préavis, son existence et sa durée résultent des usages pratiqués dans la localité et dans la profession. Un décret en Conseil d'Etat détermine les modalités d'application du présent article",
      "slug": "l1237-1",
      "source": "code_du_travail",
      "title": "L1237-1",
      "url": "https://www.legifrance.gouv.fr/affichCodeArticle.do;?idArticle=LEGIARTI000006901174&cidTexte=LEGITEXT000006072050",
    },
  ],
  "documents": Array [
    Object {
      "algo": "pre-qualified",
      "position": 0,
      "title": "Démission d'un salarié",
      "url": "/fiche-service-public/demission-dun-salarie",
    },
    Object {
      "algo": "pre-qualified",
      "position": 0,
      "title": "[DRAFT] Heures pour recherche d’emploi",
      "url": "/outils/heures-recherche-emploi",
    },
    Object {
      "algo": "pre-qualified",
      "position": 1,
      "title": "La démission : comment présenter une démission ?",
      "url": "/fiche-ministere-travail/la-demission",
    },
    Object {
      "algo": "pre-qualified",
      "position": 1,
      "title": "La démission : comment présenter une démission ?",
      "url": "/fiche-ministere-travail/la-demission#Comment-presenter-une-demission",
    },
    Object {
      "algo": "pre-qualified",
      "position": 2,
      "title": "Peut-on percevoir l'allocation chômage en cas de démission ?",
      "url": "/fiche-service-public/peut-on-percevoir-lallocation-chomage-en-cas-de-demission",
    },
    Object {
      "algo": "pre-qualified",
      "position": 3,
      "title": "La démission : faut-il respecter un préavis ?",
      "url": "/fiche-ministere-travail/la-demission#Faut-il-respecter-un-preavis",
    },
    Object {
      "algo": "pre-qualified",
      "position": 4,
      "title": "Peut-on démissionner pendant ses congés payés ?",
      "url": "/fiche-service-public/peut-on-demissionner-pendant-ses-conges-payes",
    },
    Object {
      "algo": "pre-qualified",
      "position": 5,
      "title": "Peut-on démissionner pendant un arrêt maladie ?",
      "url": "/fiche-service-public/peut-on-demissionner-pendant-un-arret-maladie",
    },
    Object {
      "algo": "pre-qualified",
      "position": 6,
      "title": "La démission : le salarié peut-il « démissionner » dans le cadre d’un contrat à durée déterminée ?",
      "url": "/fiche-ministere-travail/la-demission#Le-salarie-peut-il-lt-lt-demissionner-dans-le-cadre-d-un-contrat-a-nbsp",
    },
    Object {
      "algo": "pre-qualified",
      "position": 7,
      "title": "La démission : le refus, par le salarié, d’accepter un changement des conditions de travail est-il une démission ?",
      "url": "/fiche-ministere-travail/la-demission#Le-refus-par-le-salarie-d-accepter-un-changement-des-conditions-de-nbsp",
    },
    Object {
      "algo": "pre-qualified",
      "position": 8,
      "title": "La démission : et les heures pour recherche d’emploi ?",
      "url": "/fiche-ministere-travail/la-demission#Et-les-heures-pour-recherche-d-emploi",
    },
    Object {
      "algo": "pre-qualified",
      "position": 9,
      "title": "Peut-on démissionner pendant un congé maternité ?",
      "url": "/fiche-service-public/peut-on-demissionner-pendant-un-conge-maternite",
    },
    Object {
      "algo": "pre-qualified",
      "position": 10,
      "title": "Peut-on démissionner pendant un congé parental ?",
      "url": "/fiche-service-public/peut-on-demissionner-pendant-un-conge-parental",
    },
    Object {
      "algo": "pre-qualified",
      "position": 11,
      "title": "Démission du salarié à domicile employé par un particulier",
      "url": "/fiche-service-public/demission-du-salarie-a-domicile-employe-par-un-particulier",
    },
    Object {
      "algo": "pre-qualified",
      "position": 12,
      "title": "Démission d'une assistante maternelle",
      "url": "/fiche-service-public/demission-dune-assistante-maternelle",
    },
    Object {
      "algo": "pre-qualified",
      "position": 13,
      "title": "Article L1231-1",
      "url": "/code-du-travail/l1231-1",
    },
    Object {
      "algo": "pre-qualified",
      "position": 14,
      "title": "Article L1237-1",
      "url": "/code-du-travail/l1237-1",
    },
    Object {
      "algo": "pre-qualified",
      "position": 15,
      "title": "Article L1237-2",
      "url": "/code-du-travail/l1237-2",
    },
  ],
  "themes": Array [
    Object {
      "_score": 1.9676356,
      "algo": "fulltext",
      "cdtnId": "a9ccf1d613",
      "slug": "demission",
      "source": "themes",
      "title": "Démission",
    },
  ],
}
`;

exports[`returns unique documents from prequalified data (same slug but different source) 1`] = `
Object {
  "articles": Array [
    Object {
      "_score": 1.0930889,
      "cdtnId": "ea03713a82",
      "description": "En cas de démission, l'existence et la durée du préavis sont fixées par la loi, ou par convention ou accord collectif de travail. En l'absence de dispositions légales, de convention ou accord collectif de travail relatifs au préavis, son existence et sa durée résultent des usages pratiqués dans la localité et dans la profession. Un décret en Conseil d'Etat détermine les modalités d'application du présent article",
      "slug": "l1237-1",
      "source": "code_du_travail",
      "title": "L1237-1",
      "url": "https://www.legifrance.gouv.fr/affichCodeArticle.do;?idArticle=LEGIARTI000006901174&cidTexte=LEGITEXT000006072050",
    },
  ],
  "documents": Array [
    Object {
      "algo": "pre-qualified",
      "position": 1,
      "title": "Certificat de travail",
      "url": "/modeles-de-courriers/certificat-de-travail",
    },
    Object {
      "algo": "pre-qualified",
      "position": 2,
      "title": "/fiche-service-public/certificat-de-travail",
      "url": "/fiche-service-public/certificat-de-travail",
    },
  ],
  "themes": Array [
    Object {
      "_score": 2.648949,
      "algo": "fulltext",
      "cdtnId": "5594695039",
      "icon": "Contract",
      "slug": "embauche-et-contrat-de-travail",
      "source": "themes",
      "title": "Embauche et contrat de travail",
    },
    Object {
      "_score": 0.9986019,
      "algo": "fulltext",
      "cdtnId": "e4982f0aa9",
      "icon": "Depart",
      "slug": "depart-de-lentreprise",
      "source": "themes",
      "title": "Départ de l’entreprise",
    },
  ],
}
`;
